# CHAPTER 4-Filza Manzoor SE(A)

# PRACTICE PROBLEMS

## 4.1

Start by executing the assignment:
s = '0123456789'
Now write expressions using string s and the indexing operator that evaluate to:



(a) '234'


In [1]:
s='0123456789'
print(s[2:5])

234


(b) '78'


In [2]:
s='0123456789'
print(s[-3:-1])

78


(c) '1234567'



In [3]:
s='0123456789'
print(s[1:8])

1234567


(d) '0123'


In [4]:
s='0123456789'
print(s[1:4])

123


(e) '789' 

In [5]:
s='0123456789'
print(s[-3:])

789


## 4.2 

Assuming that variable forecast has been assigned string
'It will be a sunny day today'
write Python statements corresponding to these assignments:


(a) To variable count, the number of occurrences of string 'day' in string forecast.


In [6]:
forecast="'It will be a sunny day today'"
forecast.count('day')

2

(b) To variable weather, the index where substring 'sunny' starts.


In [7]:
forecast="'It will be a sunny day today'"
forecast.find('sunny')

14

(c) To variable change, a copy of forecast in which every occurrence of substring
'sunny' is replaced by 'cloudy'.

In [8]:
forecast="'It will be a sunny day today'"
forecast.replace('sunny','cloudy')

"'It will be a cloudy day today'"

## 4.3 

Write a statement that prints the values of variables last, first, and middle in one line,
separated by a horizontal tab character. (The Python escape sequence for the horizontal tab
character is \t.) 

In [9]:
l_name = 'Smith'
f_name = 'John'
mid_name = 'Paul'
print(l_name, f_name, mid_name, sep='\t')


Smith	John	Paul


# 4.4

Write function even() that takes a positive integer n as input and prints on the screen all
numbers between, and including, 2 and n divisible by 2 or by 3, using this output format:


In [203]:
def even():
    n=eval(input("Enter a number: "))
    for i in n:
        if i % 2==0 or i % 3==0:
            print (i)
        else:
            pass 
   
even()

Enter a number: 2,3,4,5
2
3
4


# 4.5
Assume variables first, last, street, number, city, state, zipcode have already
been assigned. Write a print statement that creates a mailing label:
John Doe
123 Main Street
AnyCity, AS 09876

In [222]:
first = 'John'
last = 'Doe'
street = 'Main Street'
number = '123'
city = 'AnyCity'
state = 'AS'
zipcode = '09876'

print(first+" "+last +"\n"+number +" "+ street + "\n"+city + "," + " "+state + " " +zipcode)

John Doe
123 Main Street
AnyCity, AS 09876


# 4.6

Implement function roster() that takes a list containing student information and prints
out a roster, as shown below. The student information, consisting of the student’s last name,
first name, class, and average course grade, will be stored in that order in a list. Therefore,
the input list is a list of lists. Make sure the roster printed out has 10 slots for every string
value and 8 for the grade, including 2 slots for the decimal part.

In [40]:
def roster(listy):
    l_names=['Joe','Mark', 'Peter', 'Kristoff','Helen']
    f_names=['Susan','Kristey','Parker','Mason','Gordon']
    grade=['sophomore','freshmen','Junior','sophomore','senior']
    avg=[ 3.45, 4.00,2.50, 2.45, 3.99]
    print('Last       First      Class       Average Grade')
    print("{:10} {:10} {:10} {:8.2f}".format(l_names[0],f_names[0],grade[0],avg[0]))
    print("{:10} {:10} {:10} {:8.2f}".format(l_names[1],f_names[1],grade[1],avg[1]))
    print("{:10} {:10} {:10} {:8.2f}".format(l_names[2],f_names[2],grade[2],avg[2]))
    print("{:10} {:10} {:10} {:8.2f}".format(l_names[3],f_names[3],grade[3],avg[3]))
    print("{:10} {:10} {:10} {:8.2f}".format(l_names[4],f_names[4],grade[4],avg[4]))
   
roster('listy')

Last       First      Class       Average Grade
Joe        Susan      sophomore      3.45
Mark       Kristey    freshmen       4.00
Peter      Parker     Junior         2.50
Kristoff   Mason      sophomore      2.45
Helen      Gordon     senior         3.99


# 4.7

Start by setting t to be the local time 1, 500, 000, 000 seconds from the start of January 1,
1970 UTC:

Construct the next strings by using the string time format function strftime():


(a) 'Thursday, July 13 2017'

In [11]:
import time
t = time.localtime()
time.strftime('%A, %B %d %Y', t)


'Tuesday, December 03 2019'

(b) '09:40 PM Central Daylight Time on 07/13/2017'

In [12]:
import time
t = time.localtime()
time.strftime('%H:%p %Z on  %m/%d/%Y', t)

'04:AM Pakistan Standard Time on  12/03/2019'

(c) 'I will meet you on Thu July 13 at 09:40 PM.'

In [13]:
import time
t = time.localtime()
time.strftime("I will meet you on "'%a %m %d on  %H:%M %p', t)

'I will meet you on Tue 12 03 on  04:28 AM'

# 4.8

Write function stringCount() that takes two string inputs—a file name and a target string—
and returns the number of occurrences of the target string in the file.


In [12]:
def stringCount(name,target):
    file=open(name,'r')
    content=file.read()
    file.close()
    count=0
    count=count+1
    return content.count(target)

stringCount('sample.txt','far')
    

2

# 4.9
Write function words() that takes one input argument—a file name—and returns the list
of actual words (without punctuation symbols !,.:;?) in the file.

In [52]:
def words(filename):
    file=open(filename,'r')
    inside= file.read()
    file.close()
    wordings=inside.split()
    return wordings
words('sample.txt')


['Far',
 'far',
 'away,',
 'behind',
 'the',
 'word',
 'mountains,',
 'far',
 'from',
 'the',
 'countries',
 'Vokalia',
 'and',
 'Consonantia,',
 'there',
 'live',
 'the',
 'blind',
 'texts.',
 'Separated',
 'they',
 'live',
 'in',
 'Bookmarksgrove',
 'right',
 'at',
 'the',
 'coast',
 'of',
 'the',
 'Semantics,',
 'a',
 'large',
 'language',
 'ocean.',
 'A',
 'small',
 'river',
 'named',
 'Duden',
 'flows',
 'by',
 'their',
 'place',
 'and',
 'supplies',
 'it',
 'with',
 'the',
 'necessary',
 'regelialia.',
 'It',
 'is',
 'a',
 'paradisematic',
 'country,',
 'in',
 'which',
 'roasted',
 'parts',
 'of',
 'sentences',
 'fly',
 'into',
 'your',
 'mouth.']

# 4.10
Implement function myGrep() that takes as input two strings, a file name and a target string,
and prints every line of the file that contains the target string as a substring.

# 4.11
theory

# EXERCISES

## 4.12

Start by running, in the interactive shell, this assignment statement:
s = 'abcdefghijklmnopqrstuvwxyz'


(a) 'bcd',


In [14]:
s ='abcdefghijklmnopqrstuvwxyz'
s[1:4]

'bcd'

(b)'abc'


In [15]:
s ='abcdefghijklmnopqrstuvwxyz'
s[0:4]

'abcd'

(c)'defghijklmnopqrstuvwx'


In [16]:
s = 'abcdefghijklmnopqrstuvwxyz'
s[3:-2]

'defghijklmnopqrstuvwx'

(d)'wxy'


In [17]:
s = 'abcdefghijklmnopqrstuvwxyz'
s[-4:-1]

'wxy'

(e)'wxyz'

In [18]:
s = 'abcdefghijklmnopqrstuvwxyz'
s[-4:]

'wxyz'

# 4.13

(a) The slice consisting of the second and third character of s is 'bc'.


In [19]:
s ='abcdefghijklmnopqrstuvwxyz'
s[1:3]

'bc'

(b) The slice consisting of the first 14 characters of s is 'abcdefghijklmn'.


In [20]:
s ='abcdefghijklmnopqrstuvwxyz'
s[0:14]

'abcdefghijklmn'

(c) The slice of s excluding the first 14 characters is 'opqrstuvwxyz'.


In [21]:
s ='abcdefghijklmnopqrstuvwxyz'
s[14:]

'opqrstuvwxyz'

(d) The slice of s excluding the first and last characters is 'bcdefghijklmnopqrstuvw'.

In [22]:
s ='abcdefghijklmnopqrstuvwxyz'
s[1:23]

'bcdefghijklmnopqrstuvw'

# 4.14

(a) Assign to variable log the next string, which happens to be a fragment of a log of a
request for a text file from a web server:
128.0.0.1 - - [12/Feb/2011:10:31:08 -0600] "GET /docs/test.txt HTTP/1.0"


In [23]:
log='128.0.0.1 - - [12/Feb/2011:10:31:08 -0600] "GET /docs/test.txt HTTP/1.0"'

(b) Assign to variable address the substring of log that ends before the first blank space
in log, using the string method split() and the indexing operator.


In [24]:
address=log.split()[0]
print (address)

128.0.0.1


(c) Assign to variable date the splice of string log containing the date (12/Feb ...
-6000), using the indexing operator on string log.

In [40]:
date=log[log.find("12/"):log.find("]")]
print(date)

12/Feb/2011:10:31:08 -0600


# 4.15


For each of the below string values of s, write the expression involving s and the string
methods split() that evaluates to list:
['10', '20', '30', '40', '50', '60']


(a) s = '10 20 30 40 50 60'



In [41]:
s = '10 20 30 40 50 60'
s.split()

['10', '20', '30', '40', '50', '60']

(b) s = '10,20,30,40,50,60'


In [54]:
s = '10,20,30,40,50,60'
s.split(sep=",")

['10', '20', '30', '40', '50', '60']

(c) s = '10&20&30&40&50&60'


In [59]:
s = '10&20&30&40&50&60'
s.split(sep='&')



['10', '20', '30', '40', '50', '60']

(d) s = '10 - 20 - 30 - 40 - 50 - 60'

In [61]:
s = '10 - 20 - 30 - 40 - 50 - 60'
s.split(sep='-')

['10 ', ' 20 ', ' 30 ', ' 40 ', ' 50 ', ' 60']

# 4.16 
Implement a program that requests three words (strings) from the user. Your program
should print Boolean value True if the words were entered in dictionary order; otherwise
nothing is printed.


In [76]:
w_1=(input("Enter the first word: "))
w_2=(input("Enter the second word: "))
w_3=(input("Enter the third word: "))
if w_1 < w_2 and w_2 < w_3:
    print (True)

Enter the first word: ali
Enter the second word: bilal
Enter the third word: sarah
True


# 4.17 
Translate each part into a Python statement using appropriate string methods:


(a) Assign to variable message the string 'The secret of this message is that
it is secret.'


In [78]:
message='The secret of this message is that it is secret.'

(b) Assign to variable length the length of string message, using operator len().


In [79]:
len(message)

48

(c) Assign to variable count the number of times the substring 'secret' appears in
string message, using string method count().


In [81]:
message.count('secret')

2

(d) Assign to variable censored a copy of string message with every occurrence of
substring 'secret' replaced by 'xxxxxx', using string method replace().

In [83]:
censored=message.replace('secret','xxxxxx')
censored

'The xxxxxx of this message is that it is xxxxxx.'

# 4.18
Suppose variable s has been assigned in this way:
s = '''It was the best of times, it was the worst of times; it
was the age of wisdom, it was the age of foolishness; it was the
epoch of belief, it was the epoch of incredulity; it was ... '''
(The beginning of A Tale of Two Cities by Charles Dickens.) Then do the following, in order:


(a) Write a sequence of statements that produce a copy of s, named newS, in which characters ., ,, ;, and \n have been replaced by blank spaces.

 

In [106]:
s = '''It was the best of times, it was the worst of times; it was the age of wisdom, it was the age of foolishness; it was the epoch of belief, it was the epoch of incredulity; it was ... '''  
tranS=s.maketrans('.,:;\n','     ')
newS=s.translate(tranS)
newS

'It was the best of times  it was the worst of times  it was the age of wisdom  it was the age of foolishness  it was the epoch of belief  it was the epoch of incredulity  it was     '

(b) Remove leading and trailing blank spaces in newS (and name the new string newS).


In [115]:
s = 'It was the best of times, it was the worst of times; it was the age of wisdom, it was the age of foolishness; it was the epoch of belief, it was the epoch of incredulity; it was ... '  
newS=s.strip()
newS

'It was the best of times, it was the worst of times; it was the age of wisdom, it was the age of foolishness; it was the epoch of belief, it was the epoch of incredulity; it was ...'

(c) Make all the characters in newS lowercase (and name the new string newS).


In [116]:
s = 'It was the best of times, it was the worst of times; it was the age of wisdom, it was the age of foolishness; it was the epoch of belief, it was the epoch of incredulity; it was ... '  
newS=s.lower()
newS

'it was the best of times, it was the worst of times; it was the age of wisdom, it was the age of foolishness; it was the epoch of belief, it was the epoch of incredulity; it was ... '

(d) Compute the number of occurrences in newS of string 'it was'.


In [117]:
s = 'It was the best of times, it was the worst of times; it was the age of wisdom, it was the age of foolishness; it was the epoch of belief, it was the epoch of incredulity; it was ... '  
newS=s.count('it was')
newS

6

(e) Change every occurrence of was to is (and name the new string newS).


In [118]:
s = 'It was the best of times, it was the worst of times; it was the age of wisdom, it was the age of foolishness; it was the epoch of belief, it was the epoch of incredulity; it was ... '  
newS=s.replace('was','is')
newS

'It is the best of times, it is the worst of times; it is the age of wisdom, it is the age of foolishness; it is the epoch of belief, it is the epoch of incredulity; it is ... '

(f) Split newS into a list of words and name the list listS.

In [121]:
s = 'It was the best of times, it was the worst of times; it was the age of wisdom, it was the age of foolishness; it was the epoch of belief, it was the epoch of incredulity; it was ... '  
newS=s.split()
listS=s.split()
print(listS)

['It', 'was', 'the', 'best', 'of', 'times,', 'it', 'was', 'the', 'worst', 'of', 'times;', 'it', 'was', 'the', 'age', 'of', 'wisdom,', 'it', 'was', 'the', 'age', 'of', 'foolishness;', 'it', 'was', 'the', 'epoch', 'of', 'belief,', 'it', 'was', 'the', 'epoch', 'of', 'incredulity;', 'it', 'was', '...']


# 4.19
Write Python statements that print the next formatted outputs using the already assigned variables first, middle, and last:

>>> first = 'Marlena'
>>> last = 'Sigel'
>>> middle = 'Mae'

(a) Sigel, Marlena Mae


In [129]:
first = 'Marlena'
last = 'Sigel'
middle = 'Mae'
print("{0} {1} {2}" .format(last,first,middle))

Sigel Marlena Mae


(b) Sigel, Marlena M.


In [131]:
print("{0}, {1} {2}." .format(last,first,middle[0:1]))

Sigel, Marlena M.


(c) Marlena M. Sigel


In [132]:
print("{0} {1}. {2}" .format(first,middle[0:1],last))

Marlena M. Sigel


(d) M. M. Sigel


In [134]:
 print("{0}. {1}. {2}" .format(first[0:1],middle[0:1],last)) 

M. M. Sigel


(e) Sigel, M.


In [137]:
 print("{1}, {0}." .format(middle[0:1],last)) 

Sigel, M.


# 4.20
Given string values for the sender, recipient, and subject of an email, write a string
format expression that uses variables sender, recipient, and subject and that prints as
shown here:
>>> sender = 'tim@abc.com'
>>> recipient = 'tom@xyz.org'
>>> subject = 'Hello!'
>>> print(???) # fill in
From: tim@abc.com
To: tom@xyz.org
Subject: Hello!


In [140]:
sender = 'tim@abc.com'
recipient = 'tom@xyz.org'
subject = 'Hello!'
print("From: {0} \nTo: {1} \nSubject: {2}". format(sender,recipient, subject )) 

From: tim@abc.com 
To: tom@xyz.org 
Subject: Hello!


# 4.21
Write Python statements that print the values of π and the Euler constant e in the
shown formats:


(a) pi = 3.1, e = 2.7


In [149]:
from math import pi
from math import e
print("pi={0:2.1f}, e={1:2.1f}".format(pi,e))

pi=3.1, e=2.7


(b) pi = 3.14, e = 2.72


In [151]:
from math import pi
from math import e
print("pi={0:3.2f}, e={1:3.2f}".format(pi,e))

pi=3.14, e=2.72


(c) pi = 3.141593e+00, e = 2.718282e+00


In [154]:
from math import pi
from math import e
print("pi={0:e}, e={1:e}".format(pi,e))

pi=3.141593e+00, e=2.718282e+00


(d) pi = 3.14159, e = 2.71828

# Problems

# 4.22
Write a function month() that takes a number between 1 and 12 as input and returns
the three-character abbreviation of the corresponding month. Do this without using an if
statement, just string operations. Hint: Use a string to store the abbreviations in order.
>>> month(1)
'Jan'
>>> month(11)
'Nov'


In [174]:
def month():
    month='JanFebMarAprMayJunJulAugSepOctNovDec'
    num=(input("Enter number between 1 to 12: "))
    if num==1:
            print (month[0:3])
    elif num==2:
            print (month[3:6])
    elif num==3:
            print (month[6:9])
    elif num==4:
            print (month[9:11])
    elif num==5:
            print (month[11:14])
    elif num==6:
            print (month[14:17]) 
    elif num==7:
            print (month[17:21])
    elif num==8:
            print (month[21:25])
    elif num==9:
            print (month[25:29])
    elif num==10:
            print (month[29:33])
    elif num==11:
            print (month[33:37])
    elif num==12:
            print (month[37:41])            
month()

Enter number between 1 to 12: 5


# 4.23
Write a function average() that takes no input but requests that the user enter a
sentence. Your function should return the average length of a word in the sentence.


In [188]:
def average():
    s = (input("Enter a sentence: "))
    words = s.split()
    return sum(len(word) for word in words)/len(words)
average()

Enter a sentence: She bought an umbrella!


5.0

# 4.24
Implement function cheer() that takes as input a team name (as a string) and prints
a cheer as shown:
>>> cheer('Huskies')
How do you spell winner?
I know, I know!
H U S K I E S !
And that's how you spell winner!
Go Huskies!


In [54]:
def cheer():
    team=input("Enter your team name: ")
    print("cheer ('{0}')". format(team))
    print("How do you spell winner? \nI know, I know!")
    print(team.upper())
    print("And that's how you spell winner! \nGo", team, "!")
cheer()

Enter your team name: huskies
cheer ('huskies')
How do you spell winner? 
I know, I know!
HUSKIES
And that's how you spell winner! 
Go huskies !


# 4.25 
Write function vowelCount() that takes a string as input and counts and prints the
number of occurrences of vowels in the string.
>>> vowelCount('Le Tour de France')
a, e, i, o, and u appear, respectively, 1, 3, 0, 1, 1 times.


In [202]:
def vowelcount():
    string=input("Enter a sentence: ")
    for i in string:
        if i=='aeiouAEIOU':
            text.count(i)
vowelcount()

Enter a sentence: Le Tour de France


# 4.26
The cryptography function crypto() takes as input a string (i.e., the name of a file in
the current directory). The function should print the file on the screen with this modification:
Every occurrence of string 'secret' in the file should be replaced with string 'xxxxxx'.
File: crypto.txt >>> crypto('crypto.txt')
I will tell you my xxxxxx. But first, I have to explain
why it is a xxxxxx.
And that is all I will tell you about my xxxxxx.


# 4.27
Write a function fcopy() that takes as input two file names (as strings) and copies
the content of the first file into the second.
File: example.txt >>> fcopy('example.txt','output.txt')
>>> open('output.txt').read()
'The 3 lines in this file end with the new line character.\n\n
There is a blank line above this line.\n' 

# 4.28 
Implement function links() that takes as input the name of an HTML file (as a
string) and returns the number of hyperlinks in that file. To do this you will assume that
each hyperlink appears in an anchor tag. You also need to know that every anchor tag ends
with the substring </a>.
Test your code on HTML file twolinks.html or any HTML file downloaded from the
web into the folder where your program is.
>>> links('twolinks.html') 
2
